### PR TITLE
TIGR-79 add filter expression engine

### PR DESCRIPTION
**Pull request by jhaynie**
**JIRA:** https://pinpt-hq.atlassian.net/browse/TIGR-79

**Description:**

Adds a filter expression engine with a custom parser grammar using PEG.

Allows us to build filters from maps similar to a very very simplified lucene syntax.

For example, we can match:

```
a: "b"
```

And that will match the following example expression:

```json
{
  "a": "b"
}
```

You can also use conjunctions like:

```
a: "b" OR a: "c"
```

### CODE REVIEW

**Comment by rblalock** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- filterexpr/filterexpr.go  2
- filterexpr/filterexpr_test.go  1
         

Clones added
============
- filterexpr/filterexpr_test.go  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Pinpoint/go-common/pullRequest?prid=5324587&bid=17424717)

**Comment by rblalock** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- filterexpr/filterexpr.go  2
- filterexpr/filterexpr_test.go  1
         

Clones added
============
- filterexpr/filterexpr_test.go  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Pinpoint/go-common/pullRequest?prid=5324587&bid=17424717)

File: Makefile [hunk 1]
```diff
@@ -6,12 +6,17 @@ ORG := pinpt
...
 
-.PHONY: all test
+.PHONY: all test gen
 
 all: test
 
 dependencies:
 	@dep ensure
+	@go get github.com/mna/pigeon
+	$(MAKE) gen
```
Comment (Robindiddams) [inline Makefile:16]: the @ will work here so we dont see our make path  `@$(MAKE) gen`

File: filterexpr/filterexpr.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package filterexpr
```
Comment (Robindiddams) [inline filterexpr/filterexpr.go:1]: I don't understand how dep is going to find the generated file when you go to import this.
I think thats what one of these is for but im not sure if that will work 🤔
```
//go:generate make gen
```

Comment (jhaynie) [reply filterexpr/filterexpr.go:1]: thats why im doing it in makefile
i could use go generate but not worth it for this case

File: filterexpr/filterexpr.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	return f.expr.Test(kv)
+}
+
+// Compile will compile the filter expression as a string in Filter which can be saved and invoked and is thread safe
+func Compile(expr string) (Filter, error) {
+	object, err := ParseReader("", strings.NewReader(expr), MaxExpressions(5000))
+	if err != nil {
+		return nil, err
+	}
+	// fmt.Println(object)
```
[the remark this answers is not in the document]
Comment (jhaynie) [reply filterexpr/filterexpr.go:30]: whats the confusion?

**Comment by rblalock** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- filterexpr/filterexpr_test.go  1
- filterexpr/filterexpr.go  2
         

Clones added
============
- filterexpr/filterexpr_test.go  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Pinpoint/go-common/pullRequest?prid=5324587&bid=17424717)

**Pull request merged**